### PR TITLE
Merge the LineFeed functionality into AdaptDispatch

### LINKED ISSUE #3673: Incorrect scroll behavior if I use minicom on Ubuntu guest
# Environment
Windows Terminal version: Version: 0.6.2951.0
Microsoft Windows [Version 10.0.18362.418]

```none
Windows build number: [run `[Environment]::OSVersion` for powershell, or `ver` for cmd]
Windows Terminal version (if applicable):

Any other software?
```

# Steps to reproduce
1. Install Ubuntu 18.04 (on WSL 1.0), install minicom inside Ubuntu.
2. Open Windows Terminal and run Ubuntu 18.04
3. Open minicom from Windows Terminal and try to capture serial console message from any embedded dev. board (say Raspberry PI)
4. You will see messages scrolling above the view port. If you scroll up you won't see the messages from the minicom.

# Expected behavior
If I scroll up, I should be able to see the serial console messages from my embedded target. 

# Actual behavior
Scroll up shows my command that I typed long ago. The serial console message is lost. If I run Ubuntu terminal directly (instead of doing the same from Windows Terminal), I could scroll up and see right messages.

Not sure, may be ncurses behavior is not compatible with Windows Terminal's expectations?

### PR DESCRIPTION
**Pull request by j4james**
The main purpose of this PR was to merge the `ITerminalApi::LineFeed`
implementations into a shared method in `AdaptDispatch`, and avoid the
VT code path depending on the `AdjustCursorPosition` function (which
could then be massively simplified). This refactoring also fixes some
bugs that existed in the original `LineFeed` implementations.

## References and Relevant Issues

This helps to close the gap between the Conhost and Terminal (#13408).
This improves some of the scrollbar mark bugs mentioned in #11000.

## Detailed Description of the Pull Request / Additional comments

I had initially hoped the line feed functionality could be implemented
entirely within `AdaptDispatch`, but there is still some Conhost and
Terminal-specific behavior that needs to be triggered when we reach the
bottom of the buffer, and the row coordinates are cycled.

In Conhost we need to trigger an accessibility scroll event, and in
Windows Terminal we need to update selection and marker offsets, reset
pattern intervals, and preserve the user's scroll offset. This is now
handled by a new `NotifyBufferRotation` method in `ITerminalApi`.

But this made me realise that the `_EraseAll` method should have been
doing the same thing when it reached the bottom of the buffer. So I've
added a call to the new `NotifyBufferRotation` API from there as well.

And in the case of Windows Terminal, the scroll offset preservation was
something that was also needed for a regular viewport pan. So I've put
that in a separate `_PreserveUserScrollOffset` method which is called
from the `SetViewportPosition` handler as well.

## Validation Steps Performed

Because of the API changes, there were a number of unit tests that
needed to be updated:

- Some of the `ScreenBufferTests` were accessing margin state in the
`SCREEN_INFORMATION` class which doesn't exist anymore, so I had to add
a little helper function which now manually detects the active margins.

- Some of the `AdapterTest` tests were dependent on APIs that no longer
exist, so they needed to be rewritten so they now check the resulting
state rather than expecting a mock API call.

- The `ScrollWithMargins` test in `ConptyRoundtripTests` was testing
functionality that didn't previously work correctly (issue #3673). Now
that it's been fixed, that test needed to be updated accordingly.

Other than getting the unit tests working, I've manually verified that
issue #3673 is now fixed. And I've also checked that the scroll markers,
selections, and user scroll offset are all being updated correctly, both
with a regular viewport pan, as well as when overrunning the buffer.

Closes #3673

### CODE REVIEW

**Comment by DHowett** [comment]
I didn't notice in the PR description that margins also moved into AdaptDispatch! Very clever, I love the new better encapsulation.

File: src/terminal/adapter/adaptDispatch.cpp [hunk 1]
```diff
@@ -2096,6 +2082,94 @@ bool AdaptDispatch::CarriageReturn()
...
+        // If the viewport has reached the end of the buffer, we can't pan down,
+        // so we cycle the row coordinates, which effectively scrolls the buffer
+        // content up. In this case we don't need to move the cursor down.
+        textBuffer.IncrementCircularBuffer(true);
+        _api.NotifyBufferRotation(1);
+
+        // We trigger a scroll rather than a redraw, since that's more efficient,
+        // but we need to turn the cursor off before doing so, otherwise a ghost
+        // cursor can be left behind in the previous position.
+        cursor.SetIsOn(false);
```
Comment (DHowett) [inline src/terminal/adapter/adaptDispatch.cpp:2155]: qq: who turns it back on?

**Review by DHowett — APPROVED** [review]
It all makes sense to me! I'd love to get @zadjii-msft's or @lhecker's hands on the ball as well.

Thanks for being so patient with us, @j4james!

File: src/terminal/adapter/adaptDispatch.cpp (hunk 1 above)
Comment (j4james) [reply src/terminal/adapter/adaptDispatch.cpp:2155]: There's a call to `_ApplyCursorMovementFlags` at the end of the method which turns the cursor back on.

**Review by zadjii-msft — APPROVED** [review]
Thank you for remedying the abomination that was `AdjustCursorPosition` - that got wildly out of hand during windows 10 and I absolutely hated every minute of it. This seems a lot easier to parse.

**Pull request merged**